### ISSUE TITLE
Desktop app

### ISSUE THREAD

**Issue by mxstbr** [withspectrum/spectrum#257]
#182 will get us quite far if implemented correctly (which I assume we will do), but having a dedicated desktop app would be really nice.

It should be possible to use Electron and to render our existing app in there, adapting a bit of code to show native notifications etc. I assume that works, but haven't tried it, so a bit of experimentation will be necessary.

**Comment by uberbryn** [comment]
OK, so this first version should just be a wrapper and push notifications.

I think if anyone has Electron experience, this is probably a good first issue.

**Comment by Raathigesh** [comment]
I shipped a couple of electron apps using https://www.electron.build/ and the experience was very pleasant. It takes care of the entire pipeline, from bundling to signing to pushing updates. https://webpack.electron.build/ is the webpack starter of electron.build.

**Comment by mxstbr** [comment]
From looking at electron.build, it seems optimized for starting a new Electron app from scratch (correct me if I'm wrong)—we need to integrate Electron into an existing app though?

**Comment by deadcoder0904** [comment]
> OK, so this first version should just be a wrapper and push notifications.

I guess you are talking about [Nativifier](https://github.com/jiahaog/nativefier) then, if its just a wrapper 😉

**Comment by spartDev** [comment]
@mxstbr like that : 
<img width="1680" alt="screen shot 2018-04-11 at 11 38 47" src="https://user-images.githubusercontent.com/531935/38608924-062b5aac-3d7d-11e8-8201-341d50c54a43.png">

**Comment by mxstbr** [comment]
Yeah like that! Except for the styling bugs :wink: PR?

**Comment by christianalfoni** [comment]
Did a ton of iterations on the Cerebral debugger related to this. Also managing auto updates. Codebase is a bit hurt by tons of iterations, but might be some helpful stuff in there.

Some takeaways:

- Get autoupdater in there
- Create a generic implementation of JSON storage, you will most likely be storing stuff and really great to have a generic way of doing it
- Create a wrapper around messaging between client and electron, giving you the possibility to wait for a task in the client. Like from the client:

```js
electron.task('whatever', 'somePayload').then(() => {
 // when task is done
})
```

It is by default just messaging between them, like `browser <=> worker`. This has a high degree of decoupling because you pass message in one place and listen for a different message to indicate it is done somewhere else. Often very nice to have it tightly coupled like the example above.

Here is that repo: https://github.com/cerebral/cerebral-debugger

Anyways, just some cents ;-)

**Comment by spartDev** [comment]
Of course about styling bugs ;)

just two thing:

1. notification does not work actually (WebSocket is closed before the connection is established)
2. When you logged in via gitHub, gitHub complain about the version of navigator:

<img width="1680" alt="screen shot 2018-04-11 at 12 14 32" src="https://user-images.githubusercontent.com/531935/38610722-fe4c03e0-3d81-11e8-82ca-67a945df9283.png">

**Comment by spartDev** [comment]
Todo:

- Packager script to ship on linux
- Packager script to ship on windows
- App Updater
- Add tray icon ?
- Add badge count & sound ?

**Comment by uberbryn** [comment]
I can take care of the icons!

**Comment by spartDev** [comment]
A thing which is on my mind, actually we load spectrum web app as a remote URL but if we want notification, badge count etc ... we need to tell  Electron app to show notification ...

Is it better to bundle the frontend into Electron with some code only for the desktop app ? What do you think ?

**Comment by mxstbr** [comment]
Yeah that's what I said in the PR, we have to bundle the frontend into Electron. :wink:

**Comment by spartDev** [comment]
OK I misunderstood that point ! 😉

**Comment by nikgraf** [comment]
btw I made good experiences using https://github.com/electron-userland/electron-forge

**Comment by spartDev** [comment]
@nikgraf electron-forge sound interesting, I would be happy to receive any ideas or suggestions.

@mxstbr any idea how to show notification count in dev ?

**Comment by mxstbr** [comment]
@spartDev I usually open Spectrum in a second, private browser window and log in with a different account, then send messages back and forth.

**Comment by uberbryn** [comment]
I like to use distinct accounts in Safari, Firefox, and Chrome for cross-browser testing.

**Comment by spartDev** [comment]
Yes, of course ! Thx!!!!

**Comment by fullstackguyhere** [comment]
Hey is there still work available for this desktop app for spectrum...? I would like to contribute!

**Comment by Gerzer** [comment]
Please consider native building Cocoa (for macOS) and .NET (Windows) apps! Electron is terrible; I **really** don’t like running separate Chromium instances for each Electron app.

**Comment by uberbryn** [comment]
@Gerzer That's not really something we can tackle right now with our limited team resources. You're welcome to dig in on that if it's a big concern of yours!

**Comment by nathankurt** [comment]
I saw this was moved to Done via automation. Does that mean this was finished, or the desktop app was abandoned?
